aidial_sdk/utils/streaming.py

```diff
@@ -60,23 +77,90 @@ async def to_block_response(stream: ResponseStream) -> dict:
 
 
 async def to_streaming_response(
-    stream: ResponseStream,
+    stream: ResponseStreamWithStr,
 ) -> AsyncGenerator[str, None]:
 
     first_chunk = await stream.__anext__()
 
     if isinstance(first_chunk, DIALException):
         raise first_chunk.to_fastapi_exception()
 
+    def _chunk_to_str(
+        chunk: Union[BaseChunkWithDefaults, DIALException, str]
+    ) -> str:
+        if isinstance(chunk, DIALException):
+            return _format_chunk(chunk.json_error())
+        elif isinstance(chunk, str):
+            return chunk
+        elif isinstance(chunk, BaseChunkWithDefaults):
+            return _format_chunk(chunk.to_dict(with_defaults=True))
+        else:
+            assert_never(chunk)
+
     async def _generator() -> AsyncGenerator[str, None]:
-        yield _format_chunk(first_chunk.to_dict(with_defaults=True))
+        yield _chunk_to_str(first_chunk)
 
         async for chunk in stream:
-            if isinstance(chunk, DIALException):
-                yield _format_chunk(chunk.json_error())
-            else:
-                yield _format_chunk(chunk.to_dict(with_defaults=True))
+            yield _chunk_to_str(chunk)
 
         yield _format_chunk(_DONE_MARKER)
 
     return _generator()
+
+
+_T = TypeVar("_T")
+
+_HeartbeatObject = Union[_T, Callable[[], Union[_T, Awaitable[_T]]]]
+_HeartbeatCallback = Callable[[], Union[None, Awaitable[None]]]
+
+
+async def _eval_heartbeat_object(o: _HeartbeatObject[_T]) -> _T:
+    if callable(o):
+        result = o()
+        if isinstance(result, Awaitable):
+            return await result
+        return cast(_T, result)
+    return o
+
+
+async def _call_heartbeat_callback(c: _HeartbeatCallback) -> None:
+    result = c()
+    if isinstance(result, Awaitable):
+        await result
+
+
+async def add_heartbeat(
+    stream: AsyncGenerator[_T, None],
+    *,
+    timeout: float,
```

Comment:
Due to function logic, that's not heartbeat timeout, that's heartbeat interval

Reply:
Indeed, fixed